server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -781,20 +777,56 @@ private void findStorageGroup(MNode node, String path, HashSet<String> res) {
    *
    * @return a list contains all distinct devices
    */
-  Set<String> getAllDevices() {
-    return new HashSet<>(getNodesList(3));
+  Set<String> getAllDevices() throws SQLException {
+    List<String> res = new ArrayList<>();
+    MNode node;
+    if ((node = getRoot()) != null) {
+      findDevices(node, "root", res);
+    }
+    return new LinkedHashSet<>(res);
+  }
+
+  private void findDevices(MNode node, String path, List<String> res) {
+    if (node == null) {
+      return;
+    }
+    if (node.isLeaf()) {
+      res.add(path);
+      return;
+    }
+    if (node.hasChildren()) {
```

Comment:
You already check the node is null or is a leaf. Here, the node should have a child. This if statement is not needed.

Reply:
Done.